add Go profiling documentation (#7197)
DataDog/documentation: content/en/tracing/profiling/_index.md

[3]: /tracing/visualization/#services
{{% /tab %}}

{{% tab "Go" %}}

The Datadog Profiler requires Go 1.12+. To begin profiling applications:

1. Get `dd-trace-go` using the command:

    ```shell

>     go get gopkg.in/DataDog/dd-trace-go.v1/ddtrace

Reviewer: ```suggestion
    go get gopkg.in/DataDog/dd-trace-go.v1/profiler
```
Would this make more sense? I'm thinking maybe yes, given step 2.

    ```

     **Note**: Profiling is available in the `dd-trace-go` library for versions 1.23.0+.

> 2. Import the [profiler][5] at the start of your application

Reviewer 2: ```suggestion
2. Import the [profiler][5] at the start of your application:
```

    ```Go
    import "gopkg.in/DataDog/dd-trace-go.v1/profiler"
    ```

> 3. To profile your code, add a [Datadog API key][1], set your envinronment service and version, then start the profiler:

Reviewer 2: ```suggestion
3. To profile your code, add a [Datadog API key][1], set your environment, service, and version, then start the profiler:
```

    ```Go
    err := profiler.Start(

>         profiler.WithAPIKey("123key")

Reviewer 2: ```suggestion
        profiler.WithAPIKey("<DATADOG_API_KEY>")
```

>         profiler.WithService("MyService"),

Reviewer 2: ```suggestion
        profiler.WithService("<SERVICE_NAME>"),
```

>         profiler.WithEnv("MyEnv"),

Reviewer 2: ```suggestion
        profiler.WithEnv("<ENVIRONMENT>"),
```

>         profiler.WithTags("version:1.2.0"),

Reviewer 2: ```suggestion
        profiler.WithTags("version:<APPLICATION_VERSION>"),
```

The tags applied are related to the profiler version or the app version ?

    )
    if err != nil {
        log.Fatal(err)
    }
    defer profiler.Stop()
    ```

> 4. After a minute or two, visualize your profiles on the [Datadog APM > Profiling page][2].

Reviewer 2: ```suggestion
4. After a minute or two, visualize your profiles in the [Datadog APM > Profiling page][2].
```

Profiler configuration:

| Environment variable                             |`profiler` method | Type          | Description                                                                                      |
| ------------------------------------------------ | ---------------- | ------------- | ------------------------------------------------------------------------------------------------ |

> |                                                  |  WithAPIKey      | String        | The Datadog [Datadog API key][1] |

Reviewer 2: ```suggestion
```

[3]: /tracing/visualization/#services
{{% /tab %}}

{{% tab "Go" %}}

The Datadog Profiler requires Go 1.12+. To begin profiling applications:

> 1. Install `ddtrace` which contains both tracing and profiling:

Reviewer: This sounds weird in terms of `go`. `ddtrace` is a package, `tracer` is another package, `profiler` is yet again another package. They aren't necessarily contained in one another. We should say something like:
```suggestion
1. Get `dd-trace-go` using the command:
```

Reviewer 3: +1

    ```shell
    go get gopkg.in/DataDog/dd-trace-go.v1/ddtrace
    ```

>      **Note**: Profiling is available in the `ddtrace` library for versions 1.23.0+.

Reviewer: Again, it's not called "`ddtrace` library". We can maybe also say:
```suggestion
     **Note**: Profiling is available in `dd-trace-go` starting version v1.23.0.
```

Reviewer 3: +1. Fwiw we had feedback from docs folks on an earlier pr that they liked formatting like `**Note**: Profiling is available in `dd-trace-go` v1.23.0+.`

Author: Done

2. Set `DD_ENV`, `DD_SERVICE`, and `DD_VERSION` in your environment variables.

> 3. Import the [Go profiler][5] at the start of your application

Reviewer: ```suggestion
3. Import the [profiler][5] at the start of your application
```
We're in the Go tab so I think that's obvious.

    ```Go
    import "gopkg.in/DataDog/dd-trace-go.v1/profiler"
    ```

4. To profile your code, add a [Datadog API key][1], then start the profiler:

    ```Go

>     err := profiler.Start(

Reviewer 3: It's odd to recommend a mix of env vars and code config. Can we choose one or the other as the primary flow and make the other secondary?

Author: Customers should be used to ENV variables on tracing.
The API key is not supprted with ENV variables today

Reviewer 3: https://docs.datadoghq.com/tracing/setup/go/#configuration shows quickstart go tracing configured via code, not env var.

        profiler.WithAPIKey("123key")
    )
    if err != nil {
        log.Fatal(err)
    }
    defer profiler.Stop()
    ```

5. After a minute or two, visualize your profiles on the [Datadog APM > Profiling page][2].

> **Note**:
> 
> - For advanced setup of the profiler or to add tags like `service` or `version`, use environment variables or `profiler` methods to set the parameters:

Reviewer 3: ```suggestion
Profiler configuration:
```

1. A single item list doesn't need to be a list :-)
2. It's funny to describe service and version tags as advanced since we (correctly) show them in the quickstart above.

| Environment variable                             |`profiler` method | Type          | Description                                                                                      |
| ------------------------------------------------ | ---------------- | ------------- | ------------------------------------------------------------------------------------------------ |
|                                                  |  WithAPIKey      | String        | The Datadog [Datadog API key][1] |
| `DD_SERVICE`                                     |  WithService     | String        | The Datadog [service][3] name, for example `my-web-app`, which can be set here, or in `DD_TAGS`.     |
| `DD_ENV`                                         |  WithEnv         | String        | The Datadog [environment][4] name, for example `production`, which can be set here, or in `DD_TAGS`. |

> | `DD_TAGS`                                        |  WithTags        | String        | The tags to apply to an uploaded profile. Must be a list of in the format `<KEY1>:<VALUE1>,<KEY2>:<VALUE2>`. |

Reviewer 3: 2 cents- I'd remove the env vars and ship this so we can start onboarding people.

Author: Why is that a problem? If possible we need to show an exhaustive list

Reviewer 3: Did we ship a release containing env vars?

Same comment from other pr- when we change interfaces (eg adding env vars) we should show what version they were changed in.

{{% /tab %}}

{{% tab "Go" %}}

Once enabled, the following profile types are collected:

| Profile type             | Definition                                                                                                                                                                                                                                                                                         |
| ------------------------ | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| CPU                      | Shows the time each function spent running on the CPU.                                                                          |

> | Allocation               | Shows the amount of heap memory allocated by each function, including allocations which were subsequently freed. Go calls this `alloc_space`              |

Reviewer 3: ```suggestion
| Allocation               | Shows the amount of heap memory allocated by each function, including allocations which were subsequently freed. Go calls this `alloc_space`. This is useful for investigating garbage collection load.              |
```

> | Allocation Count         | Shows the number of objects allocated in heap memory by each function, including allocations which were subsequently freed.     |

Reviewer 3: ```suggestion
| Allocation Count         | Shows the number of objects allocated in heap memory by each function, including allocations which were subsequently freed. This is useful for investigating garbage collection load.     |
```

> | Heap                     | Shows the amount of heap memory allocated by each function that remained allocated. Go calls this `inuse_space`.                |

Reviewer 3: ```suggestion
| Heap                     | Shows the amount of heap memory allocated by each function that remained allocated. Go calls this `inuse_space`. This is useful for investigating the overall memory usage of your service.               |
```

> | Heap Count               | Shows the number of objects allocated in heap memory by each function and that remained allocated.                              |

Reviewer 3: ```suggestion
| Heap Count               | Shows the number of objects allocated in heap memory by each function and that remained allocated. This is useful for investigating the overall memory usage of your service.                              |
```

{{% /tab %}}

{{% tab "Go" %}}

Once enabled, the following profile types are collected:

| Profile type             | Definition                                                                                                                                                                                                                                                                                         |
| ------------------------ | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| CPU                      | Shows the time each function spent running on the CPU.                                                                          |
| Allocation               | Shows the amount of heap memory allocated by each function, including allocations which were subsequently freed. Go calls this `alloc_space`              |

> | Heap                     | Shows the amount of heap memory allocated by each function that remained allocated. Go calls this `inuse_space`                 |

Reviewer 3: ![image](https://user-images.githubusercontent.com/16651846/79785148-60352600-8311-11ea-8df7-d5d8eb8afc9c.png)

Author: counts shouldn't be there, but in summary table, shouldn't they ?

Reviewer 3: Eventually, not planned for this sprint.
